### PR TITLE
Java: Enable threat models for most Java queries.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we basically opt-in all existing java queries (that are tracking dataflow from remote flow sources) to threat models.
This PR is semantics preserving in terms of results (under the *default* threat model). However, query results can change drastically if the threat model is changed (as the DCA executions below show)

There is one open question: Should we use the `ThreatModelFlowSource` class of nodes in `entrypointType` instead of `RemoteFlowSource` - this will affect all taint tracking configurations as it changes the semantics of `defaultAdditionalTaintStep`. To me it looks like we should make the change.

DCA executions show the following:
1. There are no changes in results or any indication of performance regressions on the nightly test/query suite, if the *default* threat model is applied.
2. If we extend the current threat model to also include *local* sources, thousands of new results are produced across queries for nightly/nightly. Also there doesn't appear to be any performance regressions. Some comments and explanations:
   - The query `java/concatenated-command-line` in some cases produces less results and this is because it excludes the results of `java/command-line-injection` (which might produce more results as more sources are taken into account). 
   - The query `java/tainted-arithmetic` in some cases produces lots of extra results. For the project `geogebra__geogebra` there are only two extra results. I spot checked that one of these extra results actually originated from a *local* source.
   - The query `java/path-injection` and `java/path-injection-local` produces 28 and 90 results on main and on this feature branch they produce 118 and 90 (for the project `geogebra__geogebra`), which is as expected: On main `java/path-injection-local` uses local sources and `java/path-injection` uses remote flow sources. In this feature branch, the threat model configuration, includes both the remote- and local sources and the result of `java/path-injection` is the union of the results of the `path-injection` queries on main (results have been spot checked - and the numbers add up).

### CODE REVIEW

File: java/ql/src/experimental/Security/CWE/CWE-352/JsonpInjectionLib.qll [hunk 1]
```diff
@@ -77,16 +77,26 @@ class JsonpBuilderExpr extends AddExpr {
...
+ * DEPRECATED: Use `ThreatModelFlowConfig` instead.
+ */
+deprecated module RemoteFlowConfig = ThreatModelFlowConfig;
+
+module ThreatModelFlow = DataFlow::Global<ThreatModelFlowConfig>;
+
+/**
+ * DEPRECATED: Use `ThreatModelFlow` instead.
+ */
+deprecated module RemoteFlow = ThreatModelFlow;
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-352/JsonpInjectionLib.qll:99]: This library is in `experimental` (and the path is non-importable anyway), so I'd say there's no need to deprecate these modules.

**Review by atorralba — COMMENTED** [review]
This looks great @michaelnebel, thanks for the work! I added a small comment re: deprecations.

About your question:

> Should we use the `ThreatModelFlowSource` class of nodes in `entrypointType` instead of `RemoteFlowSource`

The logic behind this predicate is that, when a source is the parameter of a method, there's good chances that some deserialization/parsing framework is involved (something that creates a Java object from input data like JSON or an HTTP form). When this happens, we expect the _whole_ object to be created from untrusted data, which means that its fields are also tainted, recursively (compared to regularly tainted objects, for which we can't say the same thing).

Now, can this be applied to local inputs? My intuition says yes, although I can think of some potential edge cases e.g. the `File` class. A tainted `File` object is a likely local source, but it contains static fields that are not under external control.

Which makes me think that, if we do this, we should exclude static fields from `entrypointFieldStep`, but otherwise seems like the right thing to do.

Another thing to keep in mind is that the additional taint steps are added to all uses of those types, not only to the source objects, which means that if local threat models include commonly-used objects, it might cause an unexpected blow up in the number of taint steps. We may need to add exclusions (like the current `TypeObject` one) if we observe this behavior.

> DCA executions show the following:

Thanks for breaking down the DCA results, they all look sensible to me, and I'm happy that everything seems to be working as expected! 😄

**Comment by michaelnebel** [comment]
Thank you very much for the review @atorralba and thank you for the background and explanation.
I suppose by excluding static fields in general, we also change the behaviour slightly, even if only remote flow sources are considered, but taken the intent of predicate into account - this seems like a reasonable change.

**Comment by atorralba** [comment]
> I suppose by excluding static fields in general, we also change the behaviour slightly, even if only remote flow sources are considered

Correct, this was a problem we already had. It's just that adding local sources to `entrypointType` could make it more frequent/noticeable, so I thought we could fix it now. Sorry for the scope creep :)

**Comment by michaelnebel** [comment]
> > I suppose by excluding static fields in general, we also change the behaviour slightly, even if only remote flow sources are considered
> 
> Correct, this was a problem we already had. It's just that adding local sources to `entrypointType` could make it more frequent/noticeable, so I thought we could fix it now. Sorry for the scope creep :)

Haha, no worries 😄 .
I will run DCA again and see if there are any changes to the alerts.
The change note still claims that we don't explicitly expect any changes to alerts. Should we modify that statement as well or is this such a corner case that we don't expect it to be noticeable?

**Comment by atorralba** [comment]
I'd be surprised if this caused changes now, but let's see what DCA has to say. If you think it's better, we can always apply that specific change in a follow-up PR.

**Comment by michaelnebel** [comment]
DCA looks good.
There are no changes to the alerts and the overall analysis time doesn't appear to be affected.
The project `geogebra__geogebra` appears to be failing for unrelated reasons.

**Pull request merged**